### PR TITLE
ANSI-589 First class confluent cloud kafka support

### PR DESCRIPTION
**Pull request by domenicbove**
# Description

Makes configuring component->ccloud kafka connection simple. All you will need are these variables: 
```
ccloud_kafka_enabled: true
ccloud_bootstrap_servers: kafka-broker1:9092
ccloud_key: client
ccloud_secret: client-secret
```
CCloud Kafka uses SASL_SSL with sasl mechanism == sasl plain, and certs signed by public cas already existing in the default java truststore. This means you *do not need a truststore* in its connection. (With the minor exception of ksqldb, more on that to come).

Changes include:
- Removes truststore properties when `ccloud_kafka_enabled` set to true 
- Skips ssl role appropriately
- Replace bootstrap.servers property with variable 
- Replaces sasl plain usernames with key/secret
- Because ksql uses `ssl.truststore.*` properties for its kafka connection and its own SSL rest end point, I added logic that downloads the ccloud root cas, and adds that to the truststore created by the ssl role. Following guide here: https://support.confluent.io/hc/en-us/articles/360040325391-How-to-use-a-custom-truststore-if-connecting-to-Confluent-Cloud
- Set `<component>_default_internal_replication_factor` to 3, this previously gets set to 1 when there is not kafka group in inventory file

Docs for component connection to CCloud Kafka
SR Docs: https://docs.confluent.io/cloud/current/cp-component/schema-reg-cloud-config.html
Connect Docs: https://docs.confluent.io/cloud/current/cp-component/connect-cloud-config.html
ksqlDB Docs: https://docs.confluent.io/cloud/current/cp-component/ksql-cloud-config.html
C3 Docs: https://docs.confluent.io/cloud/current/cp-component/c3-cloud-config.html

Notes: 
- ksqlDB docs mention, creation of ACLS. cp-ansible does not automate kafka acls, so this will not be added

TODO - Double check the rest of the properties for above components
- [x] kafka_rest
- [x] schema_registry
- [x] kafka_connect
- [x] ksql
- [x] control_center

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

`ccloud-kafka-rest` scenario built out to mock ccloud kafka cluster with sasl_ssl, puts ca into java truststore on rest of the hosts to mock out the scenario with root cas. and then creates a new ca for the component rest end points

Manually tested with real ccloud instance using first class vairables. all components worked even ksqldb- proving out the effectiveness of the ca cert scripts

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

### CODE REVIEW

File: roles/confluent.variables/defaults/main.yml [hunk 1]
```diff
@@ -868,8 +868,8 @@ sasl_plain_users: "{
     'password': 'kafka_connect-secret'
   }{% endif %}{% if 'kafka_rest' in groups %},
   'kafka_rest': {
-    'principal': 'kafka_rest',
-    'password': 'kafka_rest-secret'
+    'principal': '{% if ccloud_kafka_enabled|bool %}{{ccloud_key}}{% else %}kafka_rest{% endif %}',
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:871]: add this for each component

File: roles/confluent.kafka_rest/tasks/main.yml [hunk 2]
```diff
@@ -73,10 +73,11 @@
     cert_path: "{{kafka_rest_cert_path}}"
     key_path: "{{kafka_rest_key_path}}"
     export_certs: "{{kafka_rest_export_certs}}"
+  # CCloud Kafka does not need Keystores/Truststores
   when: >
     kafka_rest_ssl_enabled|bool or
-    kafka_broker_listeners[kafka_rest_kafka_listener_name]['ssl_enabled'] | default(ssl_enabled) | bool or
-    mds_tls_enabled | bool
+    ( kafka_broker_listeners[kafka_rest_kafka_listener_name]['ssl_enabled'] | default(ssl_enabled) | bool and not public_certificates_enabled|bool ) or
```
Comment (domenicbove) [inline roles/confluent.kafka_rest/tasks/main.yml:79]: logic missing for other components

File: roles/confluent.variables/defaults/main.yml [hunk 3]
```diff
@@ -1520,3 +1520,19 @@ control_center_pause_rolling_deployment: "{{pause_rolling_deployment}}"
 
 ### Boolean to Pause Rolling Deployment after each Kafka Connect Replicator Node starts up.
 kafka_connect_replicator_pause_rolling_deployment: "{{pause_rolling_deployment}}"
+
+
+### Boolean to configure component to Confluent Cloud Kafka. Must also set ccloud_bootstrap_servers, ccloud_key, and ccloud_key
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:1525]: There should not be a kafka or zookeeper group in your inventory file

File: roles/confluent.variables/vars/main.yml [hunk 4]
```diff
@@ -548,9 +552,11 @@ kafka_connect_properties:
...
-      producer.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']}}"
-      consumer.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']}}"
+      bootstrap.servers: "{{ ccloud_bootstrap_servers if ccloud_kafka_broker_enabled|bool else kafka_connect_bootstrap_servers }}"
+      producer.bootstrap.servers: "{{ ccloud_bootstrap_servers if ccloud_kafka_broker_enabled|bool else kafka_connect_bootstrap_servers }}"
+      # "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_connect_kafka_listener_name]['port']}}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:557]: remove commented out code

File: roles/confluent.variables/vars/main.yml [hunk 5]
```diff
@@ -733,10 +741,12 @@ ksql_properties:
       ksql.streams.producer.delivery.timeout.ms: 2147483647
       ksql.streams.producer.max.block.ms: 9223372036854775807
       listeners: "{{ksql_http_protocol}}://0.0.0.0:{{ksql_listener_port}}"
-      bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[ksql_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[ksql_kafka_listener_name]['port']}}"
+      # bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[ksql_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[ksql_kafka_listener_name]['port']}}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:744]: remove commented out code

File: roles/confluent.test/molecule/ccloud-kafka-rest/prepare.yml [hunk 6]
```diff
@@ -1,22 +1,43 @@
...
+
+- name: Kafka Rest Prepare
+  hosts: schema_registry_ccloud:kafka_connect_ccloud:kafka_rest_ccloud:ksql_ccloud:control_center_ccloud
   gather_facts: false
   tasks:
     - import_role:
-        name: confluent.kafka_broker
+        name: confluent.variables
+
+    - name: Copy CA Cert to Host
```
Comment (domenicbove) [inline roles/confluent.test/molecule/ccloud-kafka-rest/prepare.yml:33]: Better test now uses another CA for component rest end points
Might need to manually test the ksqldb

File: roles/confluent.variables/vars/main.yml [hunk 7]
```diff
@@ -1024,7 +1039,8 @@ control_center_properties:
       confluent.monitoring.interceptor.topic.replication: "{{control_center_default_internal_replication_factor}}"
       confluent.controlcenter.command.topic.replication: "{{control_center_default_internal_replication_factor}}"
       confluent.controlcenter.rest.listeners: "{{control_center_http_protocol}}://{{control_center_listener_hostname}}:{{control_center_port}}"
-      bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[control_center_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[control_center_kafka_listener_name]['port']}}"
+      # bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[control_center_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[control_center_kafka_listener_name]['port']}}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:1042]: remove commented out code

File: roles/confluent.test/molecule/ccloud/prepare.yml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    - name: Copy CA Cert to Host
+      copy:
+        src: "../../../../generated_ssl_files/{{ssl_self_signed_ca_cert_filepath|basename}}"
+        dest: /tmp/ca.crt
+
+    - name: Import CA into Java Trusted CAs
+      shell: |
+        keytool -noprompt -keystore /usr/lib/jvm/jre/lib/security/cacerts \
+          -alias CARoot \
```
Comment (domenicbove) [inline roles/confluent.test/molecule/ccloud/prepare.yml:41]: add some verify tasks

File: roles/confluent.ksql/tasks/ccloud_certs.yml [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    trust_cacert: true
+  loop:
+    - isrgrootx1
+    - isrg-root-x2
+    - lets-encrypt-r3-cross-signed
+    - lets-encrypt-r4-cross-signed
+    - lets-encrypt-e1
+    - lets-encrypt-e2
+
+# TODO remove
```
Comment (domenicbove) [inline roles/confluent.ksql/tasks/ccloud_certs.yml:29]: Remove this and update test for ksql_ssl_enabled: false

File: roles/confluent.variables/vars/main.yml [hunk 10]
```diff
@@ -897,7 +905,8 @@ kafka_rest_properties:
       ssl.keystore.password: "{{kafka_rest_keystore_storepass}}"
       ssl.key.password: "{{kafka_rest_keystore_keypass}}"
   truststore:
-    enabled: "{{kafka_rest_ssl_mutual_auth_enabled|bool or mds_tls_enabled|bool}}"
+    enabled: "{{kafka_rest_ssl_mutual_auth_enabled|bool or (mds_tls_enabled|bool and not public_certificates_enabled|bool)}}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:908]: ccloud kafka is not an mds... so technically the mds should be off here

File: VARIABLES.md [hunk 11]
```diff
@@ -1208,7 +1208,7 @@ Default:  connect-cluster
 
 Replication Factor for connect internal topics. Defaults to the minimum of the number of brokers and can be overridden via default replication factor (see default_internal_replication_factor).
 
-Default:  "{{ [ groups['kafka_broker'] | default(['localhost']) | length, default_internal_replication_factor ] | min }}"
+Default:  "{{ 3 if ccloud_kafka_broker_enabled|bool else
```
Comment (domenicbove) [inline VARIABLES.md:1211]: looks tacky

File: roles/confluent.variables/vars/main.yml [hunk 12]
```diff
@@ -936,11 +945,11 @@ kafka_rest_properties:
       confluent.monitoring.interceptor.topic: _confluent-monitoring
       producer.interceptor.classes: io.confluent.monitoring.clients.interceptor.MonitoringProducerInterceptor
       consumer.interceptor.classes: io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor
-      confluent.monitoring.interceptor.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[kafka_rest_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[kafka_rest_kafka_listener_name]['port']}}"
+      confluent.monitoring.interceptor.bootstrap.servers: "{{ kafka_rest_bootstrap_servers }}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:948]: if else

File: roles/confluent.variables/vars/main.yml [hunk 13]
```diff
@@ -797,11 +803,11 @@ ksql_properties:
       confluent.monitoring.interceptor.topic: _confluent-monitoring
       producer.interceptor.classes: io.confluent.monitoring.clients.interceptor.MonitoringProducerInterceptor
       consumer.interceptor.classes: io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor
-      confluent.monitoring.interceptor.bootstrap.servers: "{{ groups['kafka_broker'] | default(['localhost']) | join(':' + kafka_broker_listeners[ksql_kafka_listener_name]['port']|string + ',') }}:{{kafka_broker_listeners[ksql_kafka_listener_name]['port']}}"
+      confluent.monitoring.interceptor.bootstrap.servers: "{{ kafka_connect_bootstrap_servers }}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:806]: if else

File: roles/confluent.test/molecule/ccloud/molecule.yml [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        ssl_enabled: true
+        sasl_protocol: plain
+
+      all:
+        scenario_name: ccloud-kafka-rest
+
+        ssl_enabled: true
+
+        # ksql_ssl_enabled: false
+
```
Comment (domenicbove) [inline roles/confluent.test/molecule/ccloud/molecule.yml:133]: update sample inventory file

File: VARIABLES.md [hunk 15]
```diff
@@ -2956,6 +2956,30 @@ Default:  "{{pause_rolling_deployment}}"
 
 ***
 
+### ccloud_kafka_broker_enabled
+
+Boolean to configure component to Confluent Cloud Kafka. Must also set ccloud_bootstrap_servers, ccloud_key, and ccloud_key. zookeeper and kafka_broker groups should not be in inventory.
```
Comment (xnulinu) [inline VARIABLES.md:2961]:
```suggestion
Boolean to configure component to Confluent Cloud Kafka. Must also set ccloud_bootstrap_servers, ccloud_key, and ccloud_secret. zookeeper and kafka_broker groups should not be in inventory.
```

**Review by xnulinu — APPROVED** [review]
LGTM, changes look good.

**Pull request merged**